AP_GyroFFT: prevent buffer overwrite when FFT is not possible

peterbarker opened the pull request:
## Summary

Stops the autopilot boot-looping if someone fat-fingers a filter parameter.

Throws a pre-arm check if a filter is configured incorrectly.

## Description

max-Hz was being set to a value below min-Hz because of the constraint of max-Hz to be below the Nyquist limit.  This was causing start and endpoints to be reversed, leading to integer underflow leading to indexing of arrays with Vrey Large Numbers.

~4.6 version is here: https://github.com/ArduPilot/ardupilot/pull/31156~

peterbarker committed:
AP_GyroFFT: prevent buffer overwrite when FFT is not possible

max-Hz was being set to a value below min-Hz because of the constraint of max-Hz to be below the Nyquist limit.  This was causing start and endpoints to be reversed, leading to integer underflow leading to indexing of arrays with Vrey Large Numbers.

peterbarker commented:
```
Board                    AP_Periph  antennatracker  blimp  bootloader  copter  heli  iofirmware  plane  rover  sub
CubeOrange-periph-heavy  *                                 *                                                   
Durandal                            176             192    *           168     168               192    208    192
Hitec-Airspeed           *                                 *                                                   
KakuteH7-bdshot                     184             176    *           176     304               192    160    192
MatekF405                           *               *      *           *       *                 *      *      *
MatekH7A3                           192             200    *           184     192               200    200    200
Pixhawk1-1M-bdshot                  *               *                  *       *                 *      *      *
SITL_x86_64_linux_gnu               0               0                  0       0                 0      0      0
YJUAV_A6SE                          168             168    *           168     168               168    160    168
f103-QiotekPeriph        *                                 *                                                   
f303-MatekGPS            *                                 *                                                   
f303-Universal           *                                 *                                                   
iomcu                                                                                *                         
mindpx-v2                           160             168    *           160     304               168    160    32
revo-mini                           *               *      *           *       *                 *      *      *
skyviper-v2450                                                         80                                      
speedybeef4                         *               *      *           *       *                 *      *      *
```

andyp1per reviewed (changes requested):
This appears to be a redo of the previous PR which I had quite a lot of concerns over. https://github.com/ArduPilot/ardupilot/pull/31156 Has the solution changed? I also proposed a solution I would be happy with here: https://github.com/ArduPilot/ardupilot/pull/31156/changes/6754d4df5a1531e7e32f96b5edc6af7347ca7e81 I believe the changes being proposed are beyond the scope of the changes needed to prevent the issue, but happy to discuss further.

peterbarker committed:
autotest: add test for FFT prearms

peterbarker commented:
I've removed some extra code from this PR and added an autotest.

Before this fix the autotest gets this:
<img width="1768" height="1132" alt="image" src="https://github.com/user-attachments/assets/ed5f3459-2f6c-4eef-b784-aef4ffc1b743" />

Afterwards, the user gets a pre-arm failure.
```
AT-0003.5: Got: STATUSTEXT {severity : 2, text : PreArm: FFT config invalid, id : 0, chunk_seq : 0}
```

... if they fix it then the gyrofft will configure itself straight away.

peterbarker commented:
(on real hardware this is a boot-loop at start due to a hard fault.  You might be able to recover it by shaking the autopilot around to keep the accels inconsistent while you change parameters....

peterbarker commented:
> This appears to be a redo of the previous PR which I had quite a lot of concerns over. #31156

That PR was against 4.6.3 - this one was created same-day against master.  I closed that PR as being noise - things should have gone via master and then backported.

> Has the solution changed? I also proposed a solution I would be happy with here: [6754d4d](https://github.com/ArduPilot/ardupilot/commit/6754d4df5a1531e7e32f96b5edc6af7347ca7e81) I believe the changes being proposed are beyond the scope of the changes needed to prevent the issue, but happy to discuss further.

Your suggestions on that PR allowed a user to arm with a filter enabled but not functional.  A user fat-fingering a parameter could arm the vehicle without understanding that a critical filter is not functional.  That's just dangerous.
